src/Microsoft.Azure.WebJobs.Host.Storage/BlobStorageConcurrencyStatusRepository.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Blob;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    internal class BlobStorageConcurrencyStatusRepository : IConcurrencyStatusRepository
+    {
+        private const string HostContainerName = "azure-webjobs-hosts";
+        private readonly IHostIdProvider _hostIdProvider;
+        private readonly IConfiguration _configuration;
+        private readonly ILogger _logger;
+        private CloudBlobContainer _blobContainer;
+
+        public BlobStorageConcurrencyStatusRepository(IConfiguration configuration, IHostIdProvider hostIdProvider, ILoggerFactory loggerFactory)
+        {
+            _configuration = configuration;
+            _hostIdProvider = hostIdProvider;
+            _logger = loggerFactory.CreateLogger(LogCategories.Scale);
+        }
+
+        internal CloudBlobContainer Container
+        {
+            get
+            {
+                if (_blobContainer == null)
+                {
+                    string storageConnectionString = _configuration.GetWebJobsConnectionString(ConnectionStringNames.Storage);
+                    _blobContainer = CreateBlobContainer(storageConnectionString);
+                }
+                return _blobContainer;
+            }
+        }
+
+        public async Task<ConcurrencyStatusSnapshot> ReadAsync()
+        {
+            string blobPath = await GetBlobPathAsync();
+            try
+            {
+                CloudBlockBlob blob = Container.GetBlockBlobReference(blobPath);
+                if (await blob.ExistsAsync())
```

Comment:
Part of me wonders whether we need to incur the cost of checking for blob existence every time. Do we expect it to _not_ exist for long periods of time?

Reply:
Yes, good point. I can just optimistically read for the blob and handle the 404 storage exception.